cmd/prometheus/main.go

```diff
@@ -275,104 +271,182 @@ func main() {
+	{
+		g.Add(
+			func() error {
+				select {
+				case <-term:
+					level.Warn(logger).Log("msg", "Received SIGTERM, exiting gracefully...")
+				case <-waitShutdown:
+					break
+				}
+				return nil
+			},
+			func(err error) {
+			},
 		)
-		if err != nil {
-			level.Error(logger).Log("msg", "Opening storage failed", "err", err)
-			os.Exit(1)
-		}
-		level.Info(logger).Log("msg", "TSDB started")
-
-		localStorage.Set(db)
-	}()
-
-	prometheus.MustRegister(configSuccess)
-	prometheus.MustRegister(configSuccessTime)
-
-	// The notifier is a dependency of the rule manager. It has to be
-	// started before and torn down afterwards.
-	go notifier.Run()
-	defer notifier.Stop()
-
-	go ruleManager.Run()
-	defer ruleManager.Stop()
+		g.Add(
+			func() error {
+				db, err := tsdb.Open(
+					cfg.localStoragePath,
+					log.With(logger, "component", "tsdb"),
+					prometheus.DefaultRegisterer,
+					&cfg.tsdb,
+				)
+				if err != nil {
+					level.Error(logger).Log("msg", "Opening storage failed", "err", err)
+				} else {
+					localStorage.Set(db)
+				}
+				close(dbOpen)
+				<-waitShutdown
+				return nil
+			},
+			func(err error) {
+				if err := fanoutStorage.Close(); err != nil {
+					level.Error(logger).Log("msg", "Error stopping storage", "err", err)
+				} else {
+					level.Info(logger).Log("msg", "Storage closed...")
+				}
 
-	go targetManager.Run()
-	defer targetManager.Stop()
+			},
+		)
+		g.Add(
+			func() error {
+				// Wait for reload or termination signals. Start the handler for SIGHUP as
+				// early as possible, but ignore it until we are ready to handle reloading
+				// our config.
+				select {
+				case <-hupReady:
+					break
+				// in case a shutdown is initiated before the hupReady is released
+				case <-waitShutdown:
+					return nil
 
-	// Shutting down the query engine before the rule manager will cause pending queries
-	// to be canceled and ensures a quick shutdown of the rule manager.
-	defer cancelCtx()
+				}
 
-	errc := make(chan error)
-	go func() { errc <- webHandler.Run(ctx) }()
+				for {
+					select {
+					case <-hup:
+						if err := reloadConfig(cfg.configFile, logger, reloadables...); err != nil {
+							level.Error(logger).Log("msg", "Error reloading config", "err", err)
+						}
+					case rc := <-webHandler.Reload():
+						if err := reloadConfig(cfg.configFile, logger, reloadables...); err != nil {
+							level.Error(logger).Log("msg", "Error reloading config", "err", err)
+							rc <- err
+						} else {
+							rc <- nil
+						}
+					case <-waitShutdown:
+						return nil
+					}
+				}
+			},
+			func(err error) {
+				close(waitShutdown)
+			},
+		)
 
-	<-dbOpen
+		// The notifier is a dependency of the rule manager. It has to be
+		// started before and torn down afterwards.
+		g.Add(
+			func() error {
+				notifier.Run()
+				return nil
+			},
+			func(err error) {
+				notifier.Stop()
+			},
+		)
+		g.Add(
+			func() error {
+				ruleManager.Run()
+				<-waitShutdown
```

Comment:
You appear to have several actors waiting to return by receiving on the same chan. This is invalid. Each actor must have a lifecycle independent of the others.

Reply:
Oh boy , different channel for each actor...
Is this because if the actors exit at the same time the interrupt function might not run ?

Reply:
No. One actor is guaranteed to exit first. When that happens, the interrupt functions for all actors will be invoked sequentially.

Reply:
in that case why do I need a separate channel for each actor?

I will refactor some actors to be blocking and remove the `<-waitShutdown` , but others wouldn't make sence to be blocking so wanted to use a single channel for simplicity.

Some of the code in the blocks is not even actors and just some bootstrapping , but wanted to put all in groups so it is easier to follow and share the same structure (lexical scoping)

Reply:
The semantics of the Add method's parameters [are documented here](https://godoc.org/github.com/oklog/oklog/pkg/group).

> Add an actor (function) to the group. Each actor must be pre-emptable by an interrupt function. That is, [b]if interrupt is invoked, execute should return[/b]. Also, it must be safe to call interrupt even after execute has returned.

If your execute function is blocking in a chan recv on a single, shared waitForShutdown chan, then the bolded requirement is not met.

Reply:
Yes I understand the semantics,but why is this important in this implementation?  Maybe give an example worst case of what will happen if this condition is not met.

`<-waitShutdown` is there just as an ugly hack and doesn't matter if the interruptor is run after or before the executor has exited.

Reply:
or maybe give a suggestion how to improve the implementation without any major refactoring of the executors to begin with.

Reply:
You can envision the internals of the group as something like this:

1. Launch all N actor execute funcs in their own goroutines
1. Wait to receive first error from an execute func
1. For each actor
    1. Invoke interrupt func
    1. Receive error from execute func

The fact that the current version of the group code invokes all interrupt funcs first, and then waits to receive (N-1) errors second, is kind of an implementation detail, done to avoid adding per-actor tracking scaffolding. The implementation could easily change in the future. It's not safe to violate the bolded requirement; that's why it's explicitly documented as a requirement.

Reply:
got , it!

my confusion came form the current `group` implementation

this comment is a good clue about the design goals.
> // The error is passed to the interrupt functions